File: src/algorithm7/grammar.py
"""Grammar representation for head-driven parsing."""

from dataclasses import dataclass

EPSILON_TOKENS = {"ε", "eps", "epsilon"}


@dataclass
class Production:
    """A CFG production with head annotation.

    Attributes:
        lhs: Left-hand side nonterminal
        rhs: Right-hand side symbols (tuple of strings)
        head_pos: 1-indexed position of the head in rhs (τ_r in the paper)
    """
    lhs: str
    rhs: tuple[str, ...]
    head_pos: int

    def __post_init__(self):
        if not self.rhs:
            if self.head_pos != 0:
                raise ValueError("Empty RHS must use head_pos=0")
        else:
            if not 1 <= self.head_pos <= len(self.rhs):
                raise ValueError(
                    f"Head position {self.head_pos} out of range for RHS of length {len(self.rhs)}"
                )

    @property
    def head(self) -> str:
        """Return the head symbol of this production."""
        if not self.rhs:
            raise ValueError("Empty RHS has no head")
        return self.rhs[self.head_pos - 1]

    def __repr__(self) -> str:
        if not self.rhs:
            rhs_str = "ε"
        else:
            rhs_str = " ".join(
                f"[{s}]" if i == self.head_pos - 1 else s
                for i, s in enumerate(self.rhs)
            )
        return f"{self.lhs} → {rhs_str}"


class Grammar:
    """A context-free grammar with head annotations.

    Nonterminals are all symbols that appear on the left-hand side.
    The start symbol is 'S' by default.
    """

    def __init__(self, start: str = "S"):
        self.start = start
        self.productions: list[Production] = []
        self.nonterminals: set[str] = set()
        self.terminals: set[str] = set()

    def _recompute_symbols(self) -> None:
        """Recompute terminal and nonterminal sets from productions."""
        self.nonterminals = {p.lhs for p in self.productions}
        rhs_symbols = {sym for p in self.productions for sym in p.rhs}
        self.terminals = rhs_symbols - self.nonterminals

    def _unique_start_symbol(self) -> str:
        """Generate a start symbol not used elsewhere in the grammar."""
        existing = self.nonterminals | self.terminals | {self.start}
        base = f"{self.start}_START"
        candidate = base
        counter = 1
        while candidate in existing:
            candidate = f"{base}{counter}"
            counter += 1
        return candidate

    def add_production(self, lhs: str, rhs: tuple[str, ...], head_pos: int) -> int:
        """Add a production and return its index."""
        prod = Production(lhs, rhs, head_pos)
        idx = len(self.productions)
        self.productions.append(prod)
        self._recompute_symbols()
        return idx

    def is_terminal(self, symbol: str) -> bool:
        """Check if a symbol is a terminal."""
        return symbol in self.terminals

    def is_nonterminal(self, symbol: str) -> bool:
        """Check if a symbol is a nonterminal."""
        return symbol in self.nonterminals

    def productions_with_head(self, head: str) -> list[tuple[int, Production]]:
        """Return all (index, production) pairs where the head is the given symbol."""
        return [
            (i, p) for i, p in enumerate(self.productions)
            if p.head == head
        ]

    @classmethod
    def from_string(cls, text: str, start: str = "S") -> "Grammar":
        """Parse a grammar from a string representation.

        Format per line:
            LHS -> sym1 sym2 [head] sym3

        The head symbol is marked with square brackets.
        If no head is marked, the first symbol is the head.

        Example:
            S -> NP [VP]
            VP -> cl [v] NP
            NP -> [det] n

        Epsilon productions can be written as:
            A -> ε
            A -> eps
            A -> epsilon
            A ->
        """
        grammar = cls(start)
        for line in text.strip().split("\n"):
            line = line.strip()
            if not line or line.startswith("#"):
                continue

            if "->" not in line:
                raise ValueError(f"Invalid production format: {line}")

            lhs, rhs_str = line.split("->", 1)
            lhs = lhs.strip()

            rhs_parts = rhs_str.split()
            if not rhs_parts or (len(rhs_parts) == 1 and rhs_parts[0] in EPSILON_TOKENS):
                grammar.add_production(lhs, tuple(), 0)
                continue
            rhs = []
            head_pos = None

            for i, part in enumerate(rhs_parts):
                if part.startswith("[") and part.endswith("]"):
                    symbol = part[1:-1]
                    if symbol in EPSILON_TOKENS:
                        raise ValueError("Cannot mark epsilon as head")
                    head_pos = i + 1  # 1-indexed
                else:
                    symbol = part
                if symbol in EPSILON_TOKENS:
                    raise ValueError("Epsilon production must be empty (no symbols)")
                rhs.append(symbol)

            if head_pos is None:
                head_pos = 1  # Default: first symbol is head

            grammar.add_production(lhs, tuple(rhs), head_pos)

        return grammar

    def nullable_nonterminals(self) -> set[str]:
        """Compute the set of nullable nonterminals (A =>* ε)."""
        nullable = {p.lhs for p in self.productions if not p.rhs}
        changed = True
        while changed:
            changed = False
            for p in self.productions:
                if p.lhs in nullable:
                    continue
                if not p.rhs:
                    nullable.add(p.lhs)
                    changed = True
                    continue
                all_nullable = True
                for sym in p.rhs:
                    if sym not in self.nonterminals:
                        all_nullable = False
                        break
                    if sym not in nullable:
                        all_nullable = False
                        break
                if all_nullable:
                    nullable.add(p.lhs)
                    changed = True
        return nullable

    def with_single_start(self) -> "Grammar":
        """Return a grammar with a single start production.

        If multiple productions rewrite the start symbol, a fresh start
        symbol S' is introduced with a single production S' -> S.
        """
        start_prods = [p for p in self.productions if p.lhs == self.start]
        if len(start_prods) <= 1:
            return self

        new_start = self._unique_start_symbol()
        normalized = Grammar(start=new_start)
        for prod in self.productions:
            normalized.add_production(prod.lhs, prod.rhs, prod.head_pos)
        normalized.add_production(new_start, (self.start,), head_pos=1)
        return normalized

    def __repr__(self) -> str:
        lines = [f"Grammar(start={self.start!r})"]
        for i, p in enumerate(self.productions):
            lines.append(f"  [{i}] {p}")
        return "\n".join(lines)

File: ocaml/lib/grammar.ml
(** Grammar representation for head-driven parsing. *)

type symbol = string

type production = {
  lhs : symbol;
  rhs : symbol array;
  head_pos : int;
}

let head prod =
  if Array.length prod.rhs = 0 then
    invalid_arg "Empty RHS has no head"
  else
    prod.rhs.(prod.head_pos - 1)
let epsilon_tokens = ["ε"; "eps"; "epsilon"]

module StringSet = Set.Make(String)

type t = {
  start : symbol;
  mutable productions : production list;
  mutable nonterminals : StringSet.t;
  mutable terminals : StringSet.t;
}

let create ?(start = "S") () = {
  start;
  productions = [];
  nonterminals = StringSet.empty;
  terminals = StringSet.empty;
}

let recompute_symbols t =
  let nonterms =
    List.fold_left (fun acc p -> StringSet.add p.lhs acc) StringSet.empty t.productions
  in
  let rhs_symbols =
    List.fold_left (fun acc p ->
      Array.fold_left (fun acc sym -> StringSet.add sym acc) acc p.rhs
    ) StringSet.empty t.productions
  in
  t.nonterminals <- nonterms;
  t.terminals <- StringSet.diff rhs_symbols nonterms

let add_production t ~lhs ~rhs ~head_pos =
  let rhs_arr = Array.of_list rhs in
  let n = Array.length rhs_arr in
  if n = 0 then begin
    if head_pos <> 0 then
      invalid_arg "Empty RHS must use head_pos=0"
  end else if head_pos < 1 || head_pos > n then
    invalid_arg (Printf.sprintf "Head position %d out of range for RHS of length %d" head_pos n);
  let prod = { lhs; rhs = rhs_arr; head_pos } in
  let idx = List.length t.productions in
  t.productions <- t.productions @ [prod];
  recompute_symbols t;
  idx

let start t = t.start

let productions t = Array.of_list t.productions

let get_production t idx =
  let prods = productions t in
  if idx < 0 || idx >= Array.length prods then
    invalid_arg (Printf.sprintf "Production index %d out of range" idx);
  prods.(idx)

let num_productions t = List.length t.productions

let is_terminal t sym = StringSet.mem sym t.terminals

let is_nonterminal t sym = StringSet.mem sym t.nonterminals

let productions_with_head t h =
  let prods = productions t in
  Array.to_list prods
  |> List.mapi (fun i p -> (i, p))
  |> List.filter (fun (_, p) -> head p = h)

let parse_rhs_with_head parts =
  let rhs = ref [] in
  let head_pos = ref None in
  let parts = List.filter (fun s -> s <> "") parts in
  if parts = [] then
    ([], 0)
  else if List.length parts = 1 && List.mem (List.hd parts) epsilon_tokens then
    ([], 0)
  else begin
    List.iteri (fun i part ->
    let part = String.trim part in
    if String.length part >= 2 && part.[0] = '[' && part.[String.length part - 1] = ']' then begin
      let sym = String.sub part 1 (String.length part - 2) in
      if List.mem sym epsilon_tokens then
        invalid_arg "Cannot mark epsilon as head";
      head_pos := Some (i + 1);  (* 1-indexed *)
      rhs := sym :: !rhs
    end else if part <> "" then begin
      if List.mem part epsilon_tokens then
        invalid_arg "Epsilon production must be empty (no symbols)";
      rhs := part :: !rhs
    end
  ) parts;
    let rhs = List.rev !rhs in
    let head_pos = match !head_pos with
      | Some p -> p
      | None -> 1  (* Default: first symbol is head *)
    in
    (rhs, head_pos)
  end

let from_string ?(start = "S") text =
  let g = create ~start () in
  String.split_on_char '\n' text
  |> List.iter (fun line ->
       let line = String.trim line in
       if line <> "" && not (String.length line > 0 && line.[0] = '#') then
         match String.index_opt line '-' with
         | None -> failwith ("Invalid production format: " ^ line)
         | Some idx ->
           (* Check for -> *)
           if idx + 1 < String.length line && line.[idx + 1] = '>' then begin
             let lhs = String.trim (String.sub line 0 idx) in
             let rhs_str = String.sub line (idx + 2) (String.length line - idx - 2) in
             let parts = String.split_on_char ' ' (String.trim rhs_str) in
             let rhs, head_pos = parse_rhs_with_head parts in
             ignore (add_production g ~lhs ~rhs ~head_pos)
           end else
             failwith ("Invalid production format: " ^ line)
     );
  g

let pp_production fmt prod =
  if Array.length prod.rhs = 0 then
    Format.fprintf fmt "%s -> ε" prod.lhs
  else
    let rhs_strs = Array.mapi (fun i sym ->
      if i = prod.head_pos - 1 then
        "[" ^ sym ^ "]"
      else
        sym
    ) prod.rhs in
    Format.fprintf fmt "%s -> %s" prod.lhs (String.concat " " (Array.to_list rhs_strs))

let nullable_nonterminals t =
  let nullable = ref StringSet.empty in
  List.iter (fun p ->
    if Array.length p.rhs = 0 then
      nullable := StringSet.add p.lhs !nullable
  ) t.productions;
  let changed = ref true in
  while !changed do
    changed := false;
    List.iter (fun p ->
      if not (StringSet.mem p.lhs !nullable) then begin
        if Array.length p.rhs = 0 then begin
          nullable := StringSet.add p.lhs !nullable;
          changed := true
        end else begin
          let all_nullable = ref true in
          Array.iter (fun sym ->
            if not (StringSet.mem sym t.nonterminals) then
              all_nullable := false
            else if not (StringSet.mem sym !nullable) then
              all_nullable := false
          ) p.rhs;
          if !all_nullable then begin
            nullable := StringSet.add p.lhs !nullable;
            changed := true
          end
        end
      end
    ) t.productions
  done;
  StringSet.elements !nullable

let ensure_single_start t =
  let start = t.start in
  let start_count =
    List.fold_left (fun acc p -> if p.lhs = start then acc + 1 else acc) 0 t.productions
  in
  if start_count <= 1 then
    t
  else
    let symbols = StringSet.add start (StringSet.union t.nonterminals t.terminals) in
    let rec pick idx =
      let suffix = if idx = 0 then "" else string_of_int idx in
      let candidate = start ^ "_START" ^ suffix in
      if StringSet.mem candidate symbols then
        pick (idx + 1)
      else
        candidate
    in
    let new_start = pick 0 in
    let g = create ~start:new_start () in
    List.iter (fun p ->
      ignore (add_production g ~lhs:p.lhs ~rhs:(Array.to_list p.rhs) ~head_pos:p.head_pos)
    ) t.productions;
    ignore (add_production g ~lhs:new_start ~rhs:[start] ~head_pos:1);
    g

let pp fmt t =
  Format.fprintf fmt "Grammar(start=%s)@." t.start;
  let prods = productions t in
  Array.iteri (fun i p ->
    Format.fprintf fmt "  [%d] %a@." i pp_production p
  ) prods
